### PR TITLE
Add extension methods to ArmClient

### PR DESCRIPTION
**Pull request by Stolpe**
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

This solves #31579 and supercedes my previous PR (#31580), which I borked with a rebase.

A minimal example of how to use this overload can be seen below.
```csharp
const string
    subscriptionId = "",
    resourceGroupName = "",
    appServiceName = "",
    webJobName = "";

var configuration = new ConfigurationBuilder()
    .AddInMemoryCollection(new[]
    {
        KeyValuePair.Create("defaultSubscriptionId", subscriptionId)
    })
    .Build();
var services = new ServiceCollection();
services
    .AddAzureClients(builder =>
        builder.AddArmClient(configuration));
var sp = services.BuildServiceProvider();
var armClient = sp.GetRequiredService<ArmClient>();

var continuousWebJob = armClient.GetWebSiteSlotContinuousWebJobResource(
    WebSiteSlotContinuousWebJobResource.CreateResourceIdentifier(
        subscriptionId: subscriptionId,
        resourceGroupName: resourceGroupName,
        name: appServiceName,
        slot: "staging",
        webJobName: webJobName));
WebSiteSlotContinuousWebJobResource resource = await continuousWebJob.GetAsync();
```

@m-nash and @JoshLove-msft were most active in my previous PR. As requested, I have removed the constructor overload you did not like. I have added a second extension method, which takes a `defaultSubscriptionId` directly, instead of through `TConfiguration`. I hope that this can be considered good enough for both of our goals and requirements.

### CODE REVIEW

**Comment by jsquire** [comment]
@JoshLove-msft:  Would you please review?

**Review by jsquire — COMMENTED** [review]
Hi @Stolpe.  Thank you for your contribution and for helping to improve the Azure SDK experience.  

The extensions patterns look good to me.   I'll defer to @m-nash for thoughts on the ARM integration, and @JoshLove-msft as the authoritative approval for extensions patterns.

**Comment by insylogo** [comment]
Oh boy I'm excited for this !

**Review by JoshLove-msft — APPROVED** [review]
LGTM!

File: sdk/resourcemanager/Azure.ResourceManager/src/Extensions/ArmClientBuilderExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Azure.Core.Extensions;
+using Azure.ResourceManager;
+
+namespace Microsoft.Extensions.Azure
+{
+    /// <summary>
+    /// Extension methods to add <see cref="ArmClientOptions"/> client to clients builder.
```
Comment (Stolpe) [inline sdk/resourcemanager/Azure.ResourceManager/src/Extensions/ArmClientBuilderExtensions.cs:10]: I noticed that I refer to the wrong type here. When I fix it locally, and run `.\eng\scripts\CodeChecks.ps1 -ServiceDirectory resourcemanager`, I get a whole bunch of warnings. Is it safe for me to push a change that corrects this error, or will I end up messing up my PR then?
```
Validating installation instructions
git diff
diff --git a/sdk/resourcemanager/Azure.ResourceManager/src/Extensions/ArmClientBuilderExtensions.cs b/sdk/resourcemanager/Azure.ResourceManager/src/Extensions/ArmClientBuilderExtensions.cs
index 89530144e9..68d87da0d8 100644
--- a/sdk/resourcemanager/Azure.ResourceManager/src/Extensions/ArmClientBuilderExtensions.cs
+++ b/sdk/resourcemanager/Azure.ResourceManager/src/Extensions/ArmClientBuilderExtensions.cs
@@ -7,7 +7,7 @@
 namespace Microsoft.Extensions.Azure
 {
     /// <summary>
-    /// Extension methods to add <see cref="ArmClientOptions"/> client to clients builder.
+    /// Extension methods to add <see cref="ArmClient"/> client to clients builder.
     /// </summary>
     public static class ArmClientBuilderExtensions
     {
error: Generated code is not up to date.
    You may need to rebase on the latest main,
    run 'eng\scripts\Update-Snippets.ps1' if you modified sample snippets or other *.md files (https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#updating-sample-snippets),
    run 'eng\scripts\Export-API.ps1' if you changed public APIs (https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#public-api-additions).
    run 'dotnet build /t:GenerateCode' to update the generated code.
    run 'dotnet build /t:GenerateTests' to update the generated test code.

To reproduce this error locally, run 'eng\scripts\CodeChecks.ps1 -ServiceDirectory resourcemanager'.

Summary:

   1 error(s)

error : Generated code is not up to date.
    You may need to rebase on the latest main,
    run 'eng\scripts\Update-Snippets.ps1' if you modified sample snippets or other *.md files (https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#updating-sample-snippets),
    run 'eng\scripts\Export-API.ps1' if you changed public APIs (https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#public-api-additions).
    run 'dotnet build /t:GenerateCode' to update the generated code.
    run 'dotnet build /t:GenerateTests' to update the generated test code.

To reproduce this error locally, run 'eng\scripts\CodeChecks.ps1 -ServiceDirectory resourcemanager'.
```

Comment (JoshLove-msft) [reply sdk/resourcemanager/Azure.ResourceManager/src/Extensions/ArmClientBuilderExtensions.cs:10]: You will need to run `eng\scripts\Export-Api.ps1 resourcemanager`

Comment (Stolpe) [reply sdk/resourcemanager/Azure.ResourceManager/src/Extensions/ArmClientBuilderExtensions.cs]: I did that (before, and now again), and it completed successfully, without generating additional local changes:
```
PS C:\Source\per-azure-sdk-for-net> .\eng\scripts\Export-API.ps1 resourcemanager
MSBuild version 17.5.1+f6fdcf537 for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
  Azure.ResourceManager -> C:\Source\per-azure-sdk-for-net\artifacts\bin\Azure.ResourceManager\Release\netstandard2.0\Azure.ResourceManager.dll
  Generated reference assembly source code: C:\Source\per-azure-sdk-for-net\sdk\resourcemanager\Azure.ResourceManager/api/Azure.ResourceManager.netstandard2.0.cs

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:02.67
```
Still, `eng\scripts\CodeChecks.ps1 -ServiceDirectory resourcemanager` fails with the same error as before.

I created a draft PR with this fix to the main repo, and it got through all checks. This Should™️be a sound fix.

**Comment by JoshLove-msft** [comment]
I will defer to @m-nash as the owner of the resource manger library for final approval/merge.

**Comment by m-nash** [comment]
@Stolpe can we add the test cases that seem to have been dropped from the last PR https://github.com/Stolpe/azure-sdk-for-net/pull/1/files#diff-5af575d2c190ba8d297a50d4d3182af9bffb3716cb1e678668a424ce0671d579.

**Comment by Stolpe** [comment]
Of course. They are now readded, and I also added a new test to cover the endpoint that takes a `TConfiguration` parameter.

When I pushed this commit, some (old) tests started failing. I fail to see how my commit would cause these failures, and when I pushed another small commit (`var` -> `IConfiguration`), two of the four previously failed checks turned back to green. Thus, I would say that these failures are most likely due to factors not related to my change.

**Comment by m-nash** [comment]
@Stolpe one issue was your branch was far behind main and we had updated from 461 to 462.  I have pulled those into your branch and pushed.

The second issue is that when the asp tests run there is something that isn't being cleaned up and it causes all the other tests to fail but only in 462, net6 and net7 work just fine.

@JoshLove-msft can you sync this fork/branch and see if you can figure out what isn't be cleaned up in 462?

**Comment by JoshLove-msft** [comment]
> @Stolpe one issue was your branch was far behind main and we had updated from 461 to 462. I have pulled those into your branch and pushed.
> 
> The second issue is that when the asp tests run there is something that isn't being cleaned up and it causes all the other tests to fail but only in 462, net6 and net7 work just fine.
> 
> @JoshLove-msft can you sync this fork/branch and see if you can figure out what isn't be cleaned up in 462?

The ServiceProvider needs to be disposed:

```c#
private static ArmClient GetClient(ServiceCollection services)
{
    ArmClient result = null;
    ServiceProvider sp = null;
    Assert.DoesNotThrow(() =>
    {
        sp = services.BuildServiceProvider();
        result = sp
            .GetRequiredService<ArmClient>();
    });
    sp.Dispose();
    return result;
}
```

**Comment by Stolpe** [comment]
Hi,

Thanks for your help with taking my branch up to date. For the future, it seems that you merged it with Azure/azure-sdk-for-net's main. I will try to do this the next time, instead of a rebase on _my_ newly synced main.

I also added the cleanup suggested by @JoshLove-msft, though I also wrapped it into a try/finally, for the unlikely case that the Assert actually fails.

**Comment by m-nash** [comment]
Thanks again @Stolpe for the contribution the PR looks good now.

I assume you will want this shipped :-).  @ArthurMa1978 when is our next arm core release?

**Comment by Jan1503** [comment]
Ok, passing _null_ works...

**Comment by stijnherreman** [comment]
> Ok, passing _null_ works...

Passing `null` is fine, that's what some of the `ArmClient` constructors also do. But I agree it is a bit confusing.

**Pull request merged**